File: leetcode/C++/redundant-connection.cpp
class Solution {
public:
    vector<int> visit;
    int unions(int &a, int &b){
        int va=find(a), vb=find(b);
        if(va!=vb) visit[vb]=va;
        return va!=vb;
    }
    int find(int n){
        if(visit[n]==0) return n;
        return visit[n]=find(visit[n]);
    }
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        int idx=0;
        visit=vector<int>(edges.size()+1,0);
        for(int i=0; i<edges.size(); ++i){
            if(unions(edges[i][0],edges[i][1])==0) idx=i;
        }
        return edges[idx];
    }
};

File: leetcode/Python/redundant-connection.py
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        visit, idx = [0 for _ in range(len(edges)+1)], 0
        def _union(a, b):
            va, vb = _find(a), _find(b)
            if va!=vb:
                visit[va]=vb
            return va!=vb
        def _find(n):
            if visit[n]==0:
                return n
            visit[n]=_find(visit[n])
            return visit[n]
        for i, v in enumerate(edges):
            if not _union(v[0],v[1]):
                idx=i
        return edges[idx]